src/overlay/Peer.cpp

```diff
@@ -699,102 +794,168 @@ Peer::sendMessage(std::shared_ptr<StellarMessage const> msg, bool log)
+            ZoneNamedN(hmacZone, "message HMAC", true);
+            amsg.v0().sequence = self->mSendMacSeq;
+            amsg.v0().mac =
+                hmacSha256(self->mSendMacKey,
+                           xdr::xdr_to_opaque(self->mSendMacSeq.load(), *msg));
+            self->mSendMacSeq++;
+        }
+        xdr::msg_ptr xdrBytes;
+        {
+            ZoneNamedN(xdrZone, "XDR serialize", true);
+            xdrBytes = xdr::xdr_to_msg(amsg);
+        }
+        self->sendMessage(std::move(xdrBytes));
+    };
+
+    // If we're already on the background thread (i.e. via flow control), move
+    // msg to the queue right away
+    maybeExecuteInBackground("sendAuthenticatedMessage", cb);
 }
 
 bool
-Peer::isConnected() const
+Peer::isConnected(LockGuard const& stateGuard) const
 {
     return mState != CONNECTING && mState != CLOSING;
 }
 
 bool
-Peer::isAuthenticated() const
+Peer::isAuthenticated(LockGuard const& stateGuard) const
 {
     return mState == GOT_AUTH;
 }
 
+#ifdef BUILD_TESTS
+bool
+Peer::isAuthenticatedForTesting() const
+{
+    RECURSIVE_LOCK_GUARD(mStateMutex, guard);
+    return isAuthenticated(guard);
+}
+bool
+Peer::isConnectedForTesting() const
+{
+    RECURSIVE_LOCK_GUARD(mStateMutex, guard);
+    return isConnected(guard);
+}
+bool
+Peer::shouldAbortForTesting() const
+{
+    RECURSIVE_LOCK_GUARD(mStateMutex, guard);
+    return shouldAbort(guard);
+}
+#endif
+
 std::chrono::seconds
 Peer::getLifeTime() const
 {
+    releaseAssert(threadIsMain());
     return std::chrono::duration_cast<std::chrono::seconds>(
         mAppConnector.now() - mCreationTime);
 }
 
 bool
-Peer::shouldAbort() const
+Peer::shouldAbort(LockGuard const& stateGuard) const
 {
-    return mState == CLOSING || mAppConnector.overlayShuttingDown();
+    return mState == CLOSING;
 }
 
 void
 Peer::recvAuthenticatedMessage(AuthenticatedMessage&& msg)
 {
     ZoneScoped;
-    if (shouldAbort())
+    releaseAssert(!threadIsMain() || !useBackgroundThread());
+    RECURSIVE_LOCK_GUARD(mStateMutex, guard);
+
+    if (shouldAbort(guard))
     {
         return;
     }
 
-    if (mState >= GOT_HELLO && msg.v0().message.type() != ERROR_MSG)
+    std::string errorMsg;
+    if (getState(guard) >= GOT_HELLO && msg.v0().message.type() != ERROR_MSG)
     {
         if (msg.v0().sequence != mRecvMacSeq)
         {
-            ++mRecvMacSeq;
-            sendErrorAndDrop(ERR_AUTH, "unexpected auth sequence",
-                             DropMode::IGNORE_WRITE_QUEUE);
-            return;
+            errorMsg = "unexpected auth sequence";
         }
 
         if (!hmacSha256Verify(
                 msg.v0().mac, mRecvMacKey,
                 xdr::xdr_to_opaque(msg.v0().sequence, msg.v0().message)))
         {
-            ++mRecvMacSeq;
-            sendErrorAndDrop(ERR_AUTH, "unexpected MAC",
-                             DropMode::IGNORE_WRITE_QUEUE);
-            return;
+            errorMsg = "unexpected MAC";
         }
         ++mRecvMacSeq;
```

Comment:
This is also a logical race. Here's a sketch of a less-racing one:

```suggestion
        auto seq = mRecvMacKey.fetch_add(1);
        if (!hmacSha256Verify(
                msg.v0().mac, seq,
                xdr::xdr_to_opaque(msg.v0().sequence, msg.v0().message)))
        {
            errorMsg = "unexpected MAC";
        }
```